litebird_sim/mapmaking/binner.py

```diff
@@ -95,38 +109,45 @@ def _accumulate_samples_and_build_nobs_matrix(
-
-        if not additional_component:
-            # Fill the upper triangle
-            for cur_pix_idx, cur_psi in zip(pix[idet], psi[idet]):
-                cos_over_sigma = np.cos(2 * cur_psi) * inv_sigma
-                sin_over_sigma = np.sin(2 * cur_psi) * inv_sigma
-                info_pix = nobs_matrix[cur_pix_idx]
-
-                # Upper triangle
-                info_pix[0, 0] += inv_sigma2
-                info_pix[0, 1] += inv_sigma * cos_over_sigma
-                info_pix[0, 2] += inv_sigma * sin_over_sigma
-                info_pix[1, 1] += cos_over_sigma * cos_over_sigma
-                info_pix[1, 2] += sin_over_sigma * cos_over_sigma
-                info_pix[2, 2] += sin_over_sigma * sin_over_sigma
-
-        # Fill the lower triangle
-        for cur_sample, cur_pix_idx, cur_psi in zip(
-            tod[idet, :], pix[idet, :], psi[idet, :]
-        ):
-            cos_over_sigma = np.cos(2 * cur_psi) * inv_sigma
-            sin_over_sigma = np.sin(2 * cur_psi) * inv_sigma
-            info_pix = nobs_matrix[cur_pix_idx]
-
-            info_pix[1, 0] += cur_sample * inv_sigma2
-            info_pix[2, 0] += cur_sample * cos_over_sigma * inv_sigma
-            info_pix[2, 1] += cur_sample * sin_over_sigma * inv_sigma
+        if d_mask[idet]:
+            inv_sigma = 1.0 / np.sqrt(weights[idet])
+            inv_sigma2 = inv_sigma * inv_sigma
+
+            if not additional_component:
+                # Fill the upper triangle
+                for cur_pix_idx, cur_psi, cur_t_mask in zip(
+                    pix[idet], psi[idet], t_mask
+                ):
+                    if cur_t_mask:
+                        cos_over_sigma = np.cos(2 * cur_psi) * inv_sigma
+                        sin_over_sigma = np.sin(2 * cur_psi) * inv_sigma
+                        info_pix = nobs_matrix[cur_pix_idx]
+
+                        # Upper triangle
+                        info_pix[0, 0] += inv_sigma2
+                        info_pix[0, 1] += inv_sigma * cos_over_sigma
+                        info_pix[0, 2] += inv_sigma * sin_over_sigma
+                        info_pix[1, 1] += cos_over_sigma * cos_over_sigma
+                        info_pix[1, 2] += sin_over_sigma * cos_over_sigma
+                        info_pix[2, 2] += sin_over_sigma * sin_over_sigma
+
+            # Fill the lower triangle
+            for cur_sample, cur_pix_idx, cur_psi, cur_t_mask in zip(
+                tod[idet, :], pix[idet, :], psi[idet, :], t_mask
+            ):
+                if cur_t_mask:
+                    cos_over_sigma = np.cos(2 * cur_psi) * inv_sigma
+                    sin_over_sigma = np.sin(2 * cur_psi) * inv_sigma
+                    info_pix = nobs_matrix[cur_pix_idx]
+
+                    info_pix[1, 0] += cur_sample * inv_sigma2
+                    info_pix[2, 0] += cur_sample * cos_over_sigma * inv_sigma
+                    info_pix[2, 1] += cur_sample * sin_over_sigma * inv_sigma
```

Comment:
Small nitpick: with the new `if` the code gets one more nest level, but it is already quite nested with a sequence of `for`, `if`, `for`, `if`, and finally `if`. It's hard to see where the new `if d_mask[det]` ends…

A better solution would be to write the following statement at the top:

```python
if not d_mask[det]:
    continue
```

In this way it's crystal clear that if the `if` fails then the *entire* body of the `for` is skipped for this iteration.

Reply:
Do you mean something like this?
```python
if not d_mask[det]:
    continue
else:
    inv_sigma = 1.0 / np.sqrt(weights[idet])
    ....
```

Reply:
No, the `else` is redundant, the remaining part of the code can be un-indented.

Reply:
Ah I see! right!